File: leet/0199_binary_tree_right_side_view/0199_binary_tree_right_side_view.py
# LeetCode 199. Binary Tree Right Side View
# https://leetcode.com/problems/binary-tree-right-side-view/
# O(n) time, O(h) space
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        def dfs(root, depth):
            if not root:
                return
            if len(rightNodes) < depth + 1:
                rightNodes.append(root.val)
            else:
                rightNodes[depth] = root.val
            dfs(root.left, depth + 1)
            dfs(root.right, depth + 1)
        rightNodes = []
        dfs(root, 0)
        return rightNodes

File: leet/0199_binary_tree_right_side_view/0199_binary_tree_right_side_view.ts
// LeetCode 199. Binary Tree Right Side View
// https://leetcode.com/problems/binary-tree-right-side-view/
// O(n) time, O(h) space
function rightSideView(root: TreeNode | null): number[] {
    const dfs = (root, depth) => {
        if (!root)
            return;
        rightNodes[depth] = root.val;
        dfs(root.left, depth + 1);
        dfs(root.right, depth + 1);  
    };
    const rightNodes = {};
    dfs(root, 0);
    return Object.values(rightNodes);
};
